[SandboxIR] Add callbacks for instruction insert/remove/move ops

slackito opened the pull request.

slackito committed:
[SandboxIR] Add callbacks for instruction insert/remove/move ops.

tschuett commented on llvm/include/llvm/SandboxIR/Context.h:
```diff
@@ -48,6 +61,21 @@ class Context {
 
+  /// Callbacks called when an IR instruction is about to get removed. Keys are
+  /// used as IDs for deregistration.
+  DenseMap<int, RemoveInstrCallback> RemoveInstrCallbacks;
```
Do you want to keep them as ints or:
```
using ID = int;
```
?

slackito replied:
Good point, thanks! Switched it to `CallbackID`.

tschuett commented on llvm/lib/SandboxIR/Context.cpp:
```diff
@@ -660,4 +663,57 @@ Module *Context::createModule(llvm::Module *LLVMM) {
+}
+
+void Context::runInsertInstrCallbacks(Instruction *I) {
+  for (auto &CBEntry : InsertInstrCallbacks) {
```
unnecessary braces

slackito replied:
Fixed.

tschuett commented on llvm/lib/SandboxIR/Context.cpp:
```diff
@@ -660,4 +663,57 @@ Module *Context::createModule(llvm::Module *LLVMM) {
 }
 
+void Context::runRemoveInstrCallbacks(Instruction *I) {
+  for (const auto &CBEntry : RemoveInstrCallbacks) {
```
unnecessary braces

slackito replied:
Fixed.

tschuett commented on llvm/lib/SandboxIR/Context.cpp:
```diff
@@ -660,4 +663,57 @@ Module *Context::createModule(llvm::Module *LLVMM) {
+}
+
+void Context::runMoveInstrCallbacks(Instruction *I, const BBIterator &WhereIt) {
+  for (auto &CBEntry : MoveInstrCallbacks) {
```
unnecessary braces

slackito replied:
Fixed.

slackito committed:
Address some review feedback.

- Introduced `CallbackID` typedef for callback ids rather than a plain
  int.
- Remove unnecessary braces.

aeubanks commented on llvm/include/llvm/SandboxIR/Context.h:
```diff
@@ -198,6 +233,26 @@ class Context {
+  /// to be removed from its parent. Note that this will also be called when
+  /// reverting the creation of an instruction.
+  /// \Returns a callback ID for later deregistration.
+  int registerRemoveInstrCallback(RemoveInstrCallback CB);
```
```suggestion
  CallbackID registerRemoveInstrCallback(RemoveInstrCallback CB);
```
ditto below

slackito replied:
done

aeubanks commented on llvm/include/llvm/SandboxIR/Context.h:
```diff
@@ -48,6 +64,21 @@ class Context {
+  /// A counter used for assigning callback IDs during registration. The same
+  /// counter is used for all kinds of callbacks so we can detect mismatched
+  /// registration/deregistration.
+  static CallbackID NextCallbackID;
```
this shouldn't be static?

slackito replied:
Depends on whether we want callback IDs to be unique across the whole run or just for a single context object. I figured unique-per-run might be good for debugging (you never get two callback registrations with the same ID), but in practice I don't think it will make a big difference, if at all.

aeubanks replied:
I believe people will run multiple LLVM pipelines in parallel in one process, although with different LLVMContexts. so anything static will unintentionally be shared between those threads and be subject to data races

slackito replied:
Ok, that's a good reason. Made it non-static.

aeubanks commented on llvm/lib/SandboxIR/Context.cpp:
```diff
@@ -660,4 +663,54 @@ Module *Context::createModule(llvm::Module *LLVMM) {
+  return ID;
+}
+void Context::unregisterInsertInstrCallback(CallbackID ID) {
+  [[maybe_unused]] bool erased = InsertInstrCallbacks.erase(ID);
```
```suggestion
  [[maybe_unused]] bool Erased = InsertInstrCallbacks.erase(ID);
```

slackito replied:
done

aeubanks reviewed (commented):
ultra nit: no period at end of commit title

slackito committed:
Another round of feedback.

- Updated callback (de)registration method signatures in header to use
  the CallbackID type instead of int.

- Corrected case in some variable names to follow the style guide.

vporpo commented on llvm/include/llvm/SandboxIR/Context.h:
```diff
@@ -9,18 +9,34 @@
+      std::function<void(Instruction *, const BBIterator &)>;
+
+  /// An ID for a registered callback. Used for deregistration.
+  using CallbackID = int;
```
Why use an ID and not the raw pointer itself as the ID ?

tschuett replied:
non-determinism

slackito replied:
Pointers to the `std::function` objects stored in a map/vector can get invalidated when registering/unregistering more callbacks, right?

vporpo replied:
How are we going to get non-determinism ? The ID itself is not printed anywhere, it's just used as a key to remove the callback if needed. If it's the raw pointer of the function it should still work just as well. Or am I missing something?

vporpo replied:
> Pointers to the std::function objects stored in a map/vector can get invalidated when registering/unregistering more callbacks, right?

True but we can get around that in multiple ways, one of which is by allocating unique pointers and getting the raw pointer.

What I don't like about the integer IDs is that:
- they are just being used as a key and they don't really convey anything else. They are not even being used for debugging purposes.
- they need the static? counter

slackito replied:
How is another layer of indirection and an extra heap allocation better than keeping an integer around?

tschuett replied:
If you use raw pointers as keys in maps, you will get non-determinism.

The context should own the counter.

vporpo replied:
> How is another layer of indirection and an extra heap allocation better than keeping an integer around?

Yeah, it's definitely not great, and we should try to make the code that runs the callbacks as fast as possible.

Since @tschuett brought up non-determinism, should we try to make sure that the callbacks are run in the same order as registered? Because we are currently not only not getting that order but since we are iterating over the DenseMap we are also getting a non-deterministic order. Using a MapVector will fix non-determinism, but perhaps we should try to stick to registration order, and try to change it later if it turns out that it's not needed. I think we could use a vector instead of a map, because removal of a callback shouldn't happen too often (and we won't have too many callbacks) so a linear-time search should be fine. Wdyt?

slackito committed:
Make NextCallbackID not static

tmsri commented on llvm/unittests/SandboxIR/SandboxIRTest.cpp:
```diff
@@ -5962,3 +5963,87 @@ TEST_F(SandboxIRTest, CheckClassof) {
+  EXPECT_THAT(Inserted, testing::ElementsAre(NewI));
+  EXPECT_THAT(Removed, testing::IsEmpty());
+  EXPECT_THAT(Moved, testing::IsEmpty());
+
```
A drive-by comment:  How about testing insertBefore and insertAfter?

slackito replied:
Discussed this offline because @vporpo's prototype only called them on create/erase, but called them insert/remove. We only need create/erase/move callbacks for the vectorizer, so we're only adding those for now, and I have renamed them to create/erase callbacks. This leaves space to add proper insert/remove callbacks in the future if we (or some future user of Sandbox IR) needs them.

Thus, no need to test those here as they have no associated callbacks for now.

slackito committed:
Switched callbacks to MapVector to iterate in registration order.

Added test to check that registration order is respected when invoking
registered callbacks.

slackito committed:
Address more review comments, plus a couple of things discussed offline.

- Remove insert/remove callbacks to create/erase. Don't call the erase
  callback on remove. This way we have a consistent model, but we don't
  have insert/remove callbacks, just create/erase/move. The missing
  callbacks can still be added if needed in the future.

- Changed callback ids to uint64 in the unlikely case they could
  overflow a 32-bit integer in a large compile, causing hard-to-debug
  errors.

slackito committed:
clang-format

slackito commented:
Please take another look. I have also switched callback ids from `int` to `uint64_t` out of concerns (expressed offline) that we could possibly accidentally overflow a 32-bit integer (maybe registering/deregistering callbacks multiple times per region in a very large compile or something like that).

vporpo commented on llvm/lib/SandboxIR/Context.cpp:
```diff
@@ -660,4 +663,53 @@ Module *Context::createModule(llvm::Module *LLVMM) {
+
+Context::CallbackID Context::registerEraseInstrCallback(EraseInstrCallback CB) {
+  CallbackID ID = NextCallbackID++;
+  EraseInstrCallbacks[ID] = CB;
```
How about an assertion on `EraseInstrCallbacks.size()` checking that it's smaller than say 16 ? It could help catch  issues where you keep on registering callbacks by mistake.

slackito replied:
Done. I'm not super happy with having an arbitrary limit like this not documented anywhere and with a kinda low value, but it will be useful while we're developing the vectorizer and we can always revisit this decision later.

vporpo reviewed (approved):
Looks good to me. 
I think it would make sense to add some assertions on the callback map size.

slackito committed:
Add assertion for max callbacks registered at once

slackito committed:
more clang-format
